### PR TITLE
Split legacy scoring attribs into its own table

### PR DESCRIPTION
**Pull request by smoogipoo**
# Intro

As mentioned in https://github.com/ppy/osu/pull/24685#discussion_r1311208894, the current `osu_beatmap_difficulty_attribs` table presents two issues:

1. The `value` data type is a `float` which can't store total scores to the full precision required by osu! (2^31). Increasing this would significantly increase the size of data in the table.
2. ScoreV1 -> ScoreV2 requires 3 new attributes to be added for each beatmap/mod/ruleset combo. The further work presented in https://github.com/ppy/osu/pull/24166 requires at least an additional 2 or 3 attributes. This would be another significant increase in data.

# Changes

After thinking about it for a bit more, it appears like we only need to consider NoMod to get roughly accurate values. The primary reason is that `osu-queue-score-statistics` [already re-multiplies the mod multiplier](https://github.com/ppy/osu-queue-score-statistics/blob/d2b469683432a582639ff9f4de15a1edaa2bb258/osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/ImportHighScoresCommand.cs#L589-L601) since the `osu_beatmap_difficulty_attribs` table doesn't store all possible mod combinations, so it's already losing a little bit of precision.

I'm not entirely sure or confident that this theory holds up in practice, but I'm also not entirely sure the previous process did either since it hasn't been released yet. In this PR, I removed the mod multipliers throughout the simulators.

A bit of a outlier here, as you'll find reading this PR, is spinners in osu and taiko, where the number of spinner ticks depends on DT/HT being applied and the beatmap's OD. In this case, I've chosen to go with the worst-case where the number of normal ticks is maximised (bonus minimised), so that every score set can theoretically be beaten assuming enough bonus.

# Structure

The new table structure I've chosen is:

```sql
CREATE TABLE `osu_beatmap_scoring_attribs` (
    `beatmap_id` mediumint unsigned NOT NULL,
    `mode` tinyint unsigned NOT NULL,
    `legacy_accuracy_score` int NOT NULL DEFAULT 0,
    `legacy_combo_score` bigint NOT NULL DEFAULT 0,
    `legacy_bonus_score_ratio` float NOT NULL DEFAULT 0,
    PRIMARY KEY (`beatmap_id`, `mode`)
);
```

Of note is that `legacy_combo_score` is a `bigint`. The reason for this is that there are some maps that still exceed 2B score with NM (e.g. /b/1949106 - on catch).

Across 5M beatmaps, assuming 3x per-row overhead, the total table size should be ~1.5GB. As future work, we don't need to compute these attributes for unranked beatmaps, which would reduce the cost by 1 magnitude, but I need to check whether diffcalc is triggered when a beatmap is ranked without recent changes as there used to be a time where it wasn't.

# Testing

# Comparisons

osu vs master: https://docs.google.com/spreadsheets/d/1CoON2q85Ug0EO-5wQH9D1n9r2K3UpjD0O80wQeNN_qc/edit
    - I _think_ the changes here are expected. Some of it is due to the change in the spinner calcs, and the rest is how the mod multiplier is factored in.
taiko vs master: https://docs.google.com/spreadsheets/d/1CVVXz7u_KDJL6GE7BTTYlEK8gZNKUNmA4-uZ2OvvGY8/edit
    - NM and DT losses expected due to changes in swell tick counts.
catch vs master: https://docs.google.com/spreadsheets/d/1IRyONhd5naFtSGR0ciMlf_GDr3chmYwb1AV-Sx-CIlU/edit
    - Largest changes are expected due to no longer overflowing int-score.
    - Smallest changes are expected due to truncation in combo score, since mods are no longer included.
mania vs master: https://docs.google.com/spreadsheets/d/1OAggsLjX34cQY5kGSw063xHMIF_2GsJo1uqnrwTnLSA/edit
    - Changes are expected for now because mod multipliers are slightly incorrect (and were hackily correct before).

This can be tested with osu-tools via changes such as:
```diff
diff --git a/PerformanceCalculator/Performance/ScorePerformanceCommand.cs b/PerformanceCalculator/Performance/ScorePerformanceCommand.cs
index f90f2ba..8df93b3 100644
--- a/PerformanceCalculator/Performance/ScorePerformanceCommand.cs
+++ b/PerformanceCalculator/Performance/ScorePerformanceCommand.cs
@@ -4,8 +4,10 @@
 using System.Linq;
 using McMaster.Extensions.CommandLineUtils;
 using osu.Game.Beatmaps;
+using osu.Game.Database;
 using osu.Game.Models;
 using osu.Game.Online.API.Requests.Responses;
+using osu.Game.Rulesets;
 using osu.Game.Rulesets.Mods;

 namespace PerformanceCalculator.Performance
@@ -28,6 +30,9 @@ public override void Execute()

             var ruleset = LegacyHelper.GetRulesetFromLegacyID(apiScore.RulesetID);
             var score = apiScore.ToScoreInfo(apiScore.Mods.Select(m => m.ToMod(ruleset)).ToArray(), apiBeatmap);
+            score.LegacyTotalScore = score.TotalScore;
+            score.IsLegacyScore = true;
+            score.Ruleset = ruleset.RulesetInfo;
             score.BeatmapInfo.Metadata = new BeatmapMetadata
             {
                 Title = apiBeatmap.Metadata.Title,
@@ -44,6 +49,9 @@ public override void Execute()
             var performanceCalculator = ruleset.CreatePerformanceCalculator();
             var performanceAttributes = performanceCalculator?.Calculate(score, difficultyAttributes);

+            var scoreAttributes = ((ILegacyRuleset)ruleset).CreateLegacyScoreSimulator().Simulate(workingBeatmap, workingBeatmap.GetPlayableBeatmap(score.Ruleset, score.Mods));
+            var totalScore = StandardisedScoreMigrationTools.ConvertFromLegacyTotalScore(score, scoreAttributes);
+
             OutputPerformance(score, performanceAttributes, difficultyAttributes);
         }
     }
```

### CODE REVIEW

**Comment by bdach** [comment]
> Some of it is due to the change in the spinner calcs

Can you elaborate which "change" this is referring to, because I'm having trouble recalling any? The two spinner changes we have that I immediately thought of (https://github.com/ppy/osu/pull/24661, https://github.com/ppy/osu/pull/24662) are both unmerged yet.

> This can be tested with osu-tools via changes such as:

Is there anything in particular that I should be testing?

File: osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs [hunk 1]
```diff
@@ -133,22 +129,25 @@ private void simulateHit(HitObject hitObject)
                     // The spinner object applies a lenience because gameplay mechanics differ from osu-stable.
                     // We'll redo the calculations to match osu-stable here...
                     const double maximum_rotations_per_second = 477.0 / 60;
-                    double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(playableBeatmap.Difficulty.OverallDifficulty, 3, 5, 7.5);
+
+                    // Normally, this value depends on the final overall difficulty. For simplicity, we'll only consider the worst case that minimises bonus score.
+                    const double minimum_rotations_per_second = 7.5;
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs:134]: I'm trying to make heads and tails of this and I'm not sure it's making sense to me. The thought process is as follows:

- We are assuming the maximum... minimum of number of rotations per second here, OD-wise. This means that the rotation requirement is going to be as high as it can possibly can go here.
- Therefore, setting this to max also maximises `halfSpinsRequiredForCompletion`.
- That indeed does mean that the total final achievable legacy bonus score for spinners will be _minimised_.
- However, after this I'm lost, because what happens later is that `BonusScoreRatio` is computed as `standardisedBonusScore / legacyBonusScore`. If the denominator is _minimised_, the quotient itself is _maximised_.
- The score migrator uses `BonusScoreRatio` by taking the legacy total score, by taking the total score of a given stable score, subtracting the max legacy base score, and if there is any excess left, then that is considered bonus and is multiplied by `BonusScoreRatio`. But as stated above, the ratio is _maximised_ here, and the OP says:

> where the number of normal ticks is maximised (bonus minimised), so that every score set can theoretically be beaten assuming enough bonus.

and is where I get lost, because it seems to me that this _benefits_ legacy scores...? Where am I wrong?

Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs:134]: The above does seem to check out.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs:134]: Hmmm... I guess one of the biggest issues is that I thought normal ticks were part of `AccuracyScore`, but this isn't the case - they're actually bonus.

Excluding that, my thought process here (if normal ticks were part of `AccuracyScore`) was to maximise `(AccuracyScore + ComboScore)` thereby minimising the calculation of the non-bonus score part when converting high scores.  
In that scenario, if the only bonus was from `SpinnerBonusTick`, `BonusScoreRatio` is actually a straight up constant equal to `50 / 1100` so there's no such maximisation/minimisation of the denominator.  

But since normal ticks are bonus, then I believe it is true that we want to minimise the number of normal ticks as they're the dominator of the equation.

We may also want to consider removing `BonusScoreRatio` and treating it as a constant of `0.045`.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs]: Right, everything you said above checks out to me :+1:

File: osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs [hunk 2]
```diff
@@ -133,22 +129,25 @@ private void simulateHit(HitObject hitObject)
                     // The spinner object applies a lenience because gameplay mechanics differ from osu-stable.
                     // We'll redo the calculations to match osu-stable here...
                     const double maximum_rotations_per_second = 477.0 / 60;
-                    double minimumRotationsPerSecond = IBeatmapDifficultyInfo.DifficultyRange(playableBeatmap.Difficulty.OverallDifficulty, 3, 5, 7.5);
+
+                    // Normally, this value depends on the final overall difficulty. For simplicity, we'll only consider the worst case that maximises bonus score.
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/OsuLegacyScoreSimulator.cs:133]: If anything I'd maybe consider expanding this with an explanation of _why_ we're maximising bonus score. Something like

```suggestion
                    // Normally, this value depends on the final overall difficulty. For simplicity, we'll only consider the worst case that maximises bonus score.
                    // As we're primarily concerned with computing the maximum theoretical final score,
                    // this will have the final effect of slightly underestimating bonus score achieved on stable when converting from score V1.
```

perhaps?

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs [hunk 3]
```diff
@@ -109,21 +103,24 @@ private void simulateHit(HitObject hitObject)
...
-                        halfSpinsRequiredForCompletion = Math.Max(1, (int)(halfSpinsRequiredForCompletion * 0.75f));
-                    if (mods.Any(m => m is ModHalfTime))
-                        halfSpinsRequiredForCompletion = Math.Max(1, (int)(halfSpinsRequiredForCompletion * 1.5f));
+                    //
+                    // Normally, this multiplier depends on the active mods (DT = 0.75, HT = 1.5). For simplicity, we'll only consider the worst case that maximises rotations.
+                    // This way, scores remain beatable at the cost of the conversion being slightly inaccurate.
+                    //   - A perfect DT/NM score will have less than 1M total score (excluding bonus).
+                    //   - A perfect HT score will have 1M total score (excluding bonus).
+                    //
+                    halfSpinsRequiredForCompletion = Math.Max(1, (int)(halfSpinsRequiredForCompletion * 1.5f));
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:120]: Hmm, on a second check, aren't changes similar to https://github.com/ppy/osu/pull/24779/commits/ed295effc5206fcb347d2e73cadc5c5abb9251db required here too?

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:120]: Unless I'm missing something, I don't believe so because swell ticks are base score not bonus?

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:120]: Hmm ok, that they are. My bad.

**Review by bdach — DISMISSED** [review]
I think this is okay, but probably not gonna click buttons before @peppy has a chance to look this over.

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs [hunk 4]
```diff
@@ -159,8 +156,7 @@ private void simulateHit(HitObject hitObject)
             {
                 int oldScoreIncrease = scoreIncrease;
 
-                // ReSharper disable once PossibleLossOfFraction (intentional to match osu-stable...)
-                scoreIncrease += (int)(scoreIncrease / 35 * 2 * (difficultyPeppyStars + 1) * modMultiplier) * (Math.Min(100, combo) / 10);
```
Comment (peppy) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Was this change intentional? It looks like it's undoing something specifically put in place to match expectations.

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Also, while we're at it, why is `modMultiplier` just *gone* from this line, too?

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Unless I'm missing something, as far as I can tell, the only thing that's changed is the removal of `modMultiplier`? This is explained by removing the mod multiplier as I mentioned in the description.

I didn't explicitly mention it, but I thought that was a given, given the entire purpose here is to remove any mod multipliers.

The same is done for osu!, just via removal of it being multiplied into `scoreMultiplier` (https://github.com/ppy/osu/pull/24779/files#diff-cb9b51e6336d258f91476bf755934d4e2e1593f9b697cfa7ea7a05f6886b4e0fL76-R66).

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Okay, fair point on the multiplier disappearing, but the other thing that's changed is that the `35` factor became a `35f`, therefore implicitly performing a promotion to float, while it was previously *intentionally* doing integer division with a comment suppression. So that is something that may have changed values unintentionally.

Comment (peppy) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Maybe `ILegacyScoreSimulator.Simulate`'s xmldoc should be updated to mention that it doesn't apply any mod multipliers and that it should be done in post? Probably as a remark.

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: Hmm okay I was the impression that the `double` being there would cause an implicit conversion of int-to-double before the operations.

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: > the `double` being there would cause an implicit conversion of int-to-double before the operations

It would, but my point is that that is probably undesirable. For reference:

image

https://github.com/peppy/osu-stable-reference/blob/1531237b63392e82c003c712faa028406073aa8f/osu!/GameModes/Play/Rulesets/Ruleset.cs#L1093-L1098

Comment (smoogipoo) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoLegacyScoreSimulator.cs:163]: I mean that the double was there as part of `modMultiplier` previously, so I was trying to preserve behaviour, but that is incorrect.

**Review by bdach — APPROVED** [review]
went over this again and i _think_ there should be no further surprises this time

**Review by peppy — APPROVED** [review]
